### PR TITLE
GH-41133: [Benchmarking] Build benchmarks in benchmarks.env

### PR DESCRIPTION
**Pull request by austin3dickey**
Previously, we were relying on `archery` to build C++ microbenchmarks during Conbench runs. That doubled the build time and ignored all the environment variables set in this file. Along with https://github.com/voltrondata-labs/benchmarks/pull/159, this change will remove the duplicated build.
* GitHub Issue: #41133

### CODE REVIEW

**Comment by austin3dickey** [comment]
@ursabot please benchmark lang=C++

**Comment by austin3dickey** [comment]
@ursabot please benchmark lang=C++

**Comment by austin3dickey** [comment]
@ursabot please benchmark lang=C++

**Comment by austin3dickey** [comment]
@ursabot please benchmark lang=C++

File: dev/archery/archery/benchmark/google.py [hunk 1]
```diff
@@ -57,7 +57,7 @@ def results(self, repetitions=1, repetition_min_time=None):
                     "--benchmark_out_format=json"]
 
             if repetition_min_time is not None:
-                argv.append(f"--benchmark_min_time={repetition_min_time:.6f}")
+                argv.append(f"--benchmark_min_time={repetition_min_time:.6f}s")
```
Comment (austin3dickey) [inline dev/archery/archery/benchmark/google.py:60]: This was to suppress a message I saw in the logs:
```
LOG(0): Value passed to --benchmark_min_time should have a suffix. Eg., `30s` for 30-seconds.
```

File: dev/conbench_envs/benchmarks.env [hunk 2]
```diff
@@ -27,6 +28,7 @@ ARROW_GANDIVA=OFF
 ARROW_HDFS=ON
 ARROW_HOME=$CONDA_PREFIX
 ARROW_INSTALL_NAME_RPATH=ON
+ARROW_JEMALLOC=OFF
```
Comment (austin3dickey) [inline dev/conbench_envs/benchmarks.env:31]: This used to be set [here](https://github.com/voltrondata-labs/arrow-benchmarks-ci/commit/7c20a64e464705d5bf4e2f20f4c8befd601d566a) but I think it belongs in this file instead.

File: dev/archery/archery/benchmark/google.py (hunk 1 above)
Comment (pitrou) [reply dev/archery/archery/benchmark/google.py:60]: Yes, it's a change in the Google benchmark CLI.

**Comment by pitrou** [comment]
> I think the regressions make sense because we're now building Arrow differently, with different cmake options. This should be fine and is a reflection of the environment setup, not the code.

This sounds fine to me.

**Comment by austin3dickey** [comment]
For reference, here's the full diff of cmake flags:

<details>

```
> diff before.txt after.txt
10c10
<    ARROW_GIT_ID=835e21873522bba2928beb972e863790f4c270cb [default=""]
---
>    ARROW_GIT_ID=615e38fe479e14a7f270a6a1eff7de2e9bf94cc0 [default=""]
12c12
<    ARROW_GIT_DESCRIPTION=apache-arrow-16.0.0.dev-483-g835e21873 [default=""]
---
>    ARROW_GIT_DESCRIPTION=apache-arrow-16.0.0.dev-422-g615e38fe4 [default=""]
14c14
<    ARROW_NO_DEPRECATED_API=OFF [default=OFF]
---
>    ARROW_NO_DEPRECATED_API=ON [default=OFF]
23c23
<    ARROW_USE_LD_GOLD=ON [default=OFF]
---
>    ARROW_USE_LD_GOLD=OFF [default=OFF]
25,26d24
<    ARROW_USE_LLD=OFF [default=OFF]
<        Use the LLVM lld for linking (if available)
60c58
<    ARROW_BUILD_BENCHMARKS_REFERENCE=OFF [default=OFF]
---
>    ARROW_BUILD_BENCHMARKS_REFERENCE=ON [default=OFF]
99c97
<    ARROW_BUILD_UTILITIES=OFF [default=OFF]
---
>    ARROW_BUILD_UTILITIES=ON [default=OFF]
119c117
<    ARROW_HDFS=OFF [default=OFF]
---
>    ARROW_HDFS=ON [default=OFF]
127c125
<    ARROW_MIMALLOC=OFF [default=OFF]
---
>    ARROW_MIMALLOC=ON [default=OFF]
131c129
<    ARROW_ORC=OFF [default=OFF]
---
>    ARROW_ORC=ON [default=OFF]
136c134
<    ARROW_S3=OFF [default=OFF]
---
>    ARROW_S3=ON [default=OFF]
149c147
<    ARROW_DEPENDENCY_SOURCE=CONDA [default=CONDA|AUTO|BUNDLED|SYSTEM|VCPKG|BREW]
---
>    ARROW_DEPENDENCY_SOURCE=AUTO [default=CONDA|AUTO|BUNDLED|SYSTEM|VCPKG|BREW]
216c214
<    PARQUET_BUILD_EXECUTABLES=OFF [default=OFF]
---
>    PARQUET_BUILD_EXECUTABLES=ON [default=OFF]
218c216
<    PARQUET_BUILD_EXAMPLES=OFF [default=OFF]
---
>    PARQUET_BUILD_EXAMPLES=ON [default=OFF]
220c218
<    PARQUET_REQUIRE_ENCRYPTION=OFF [default=OFF]
---
>    PARQUET_REQUIRE_ENCRYPTION=ON [default=OFF]
228c226
<    ARROW_GANDIVA_PC_CXX_FLAGS="" [default=""]
---
>    ARROW_GANDIVA_PC_CXX_FLAGS=-isystem;/var/lib/buildkite-agent/.conda/envs/arrow-commit/bin/../lib/gcc/aarch64-conda-linux-gnu/12.3.0/include;-isystem;/var/lib/buildkite-agent/.conda/envs/arrow-commit/bin/../lib/gcc/aarch64-conda-linux-gnu/12.3.0/include-fixed;-isystem;/var/lib/buildkite-agent/.conda/envs/arrow-commit/bin/../lib/gcc/aarch64-conda-linux-gnu/12.3.0/../../../../aarch64-conda-linux-gnu/include;-isystem;/var/lib/buildkite-agent/.conda/envs/arrow-commit/bin/../lib/gcc/../../aarch64-conda-linux-gnu/include/c++/12.3.0;-isystem;/var/lib/buildkite-agent/.conda/envs/arrow-commit/bin/../lib/gcc/../../aarch64-conda-linux-gnu/include/c++/12.3.0/aarch64-conda-linux-gnu;-isystem;/var/lib/buildkite-agent/.conda/envs/arrow-commit/bin/../lib/gcc/../../aarch64-conda-linux-gnu/include/c++/12.3.0/backward;-isystem;/var/lib/buildkite-agent/.conda/envs/arrow-commit/bin/../aarch64-conda-linux-gnu/sysroot/usr/include; [default=""]
```
</details>

**Comment by pitrou** [comment]
I think you can disable `ARROW_BUILD_BENCHMARKS_REFERENCE` otherwise you'll get more unstable benchmark results. You could probably also disable the examples, executables, utilities, as well as S3 and ORC.

**Comment by austin3dickey** [comment]
>I think you can disable...

Cool, I'll try that out. Note that with this change we started running these additional suites. Not sure exactly what flags turned them on, but it's probably nice to run them if we can.
- arrow-range-benchmark
- arrow-compression-benchmark
- arrow-machine-benchmark

**Comment by austin3dickey** [comment]
@ursabot please benchmark lang=C++

File: dev/archery/archery/benchmark/google.py (hunk 1 above)
Comment (assignUser) [reply dev/archery/archery/benchmark/google.py:60]: It seems that the warning was added with the flag itself https://github.com/google/benchmark/blame/70916cbf71f50b9e1e6f13559e10d6dbb92beb32/src/benchmark_runner.cc#L199 so this isn't an issue after all

**Review by assignUser — APPROVED** [review]
LGTM

**Comment by assignUser** [comment]
Though it is not a MINOR, I have created an issue. @austin3dickey please comment 'Take' on it to get it assigned to you (gh anti-spam prevents me from assigning you if you haven't interacted with it)

**Pull request merged**